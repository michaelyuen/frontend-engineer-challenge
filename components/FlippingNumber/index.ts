export * from "./FlippingNumber";
export * from "./styles";
