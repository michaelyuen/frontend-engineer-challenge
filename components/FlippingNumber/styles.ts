import styled, { css, keyframes } from "styled-components";

const flip = keyframes`
  0% {
    transform: rotateX(0);
  }

  100% {
    border-bottom: 1px solid black;
    transform: rotateX(-0.5turn);
    z-index: 1;
  }
`;

const show = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

export const FlippingNumberContainer = styled.div`
  position: relative;
  text-align: center;
  z-index: 1;

  p {
    font-weight: bold;
    letter-spacing: 5px;
    margin-top: 1em;
    text-transform: uppercase;
  }
`;

export interface NumberContainerProps {
  animationDuration: number;
  shouldAnimate: boolean;
}

export const NumberContainer = styled.div`
  color: var(--flipping-number-color);
  font-size: 70px;
  height: 100px;
  line-height: 100px;
  perspective: 500px;
  position: relative;

  @media (min-width: 768px) {
    font-size: 100px;
    height: 150px;
    line-height: 150px;
  }

  div {
    display: flex;
    justify-content: center;
    left: 0;
    position: absolute;
    width: 100%;

    &:not(.flipper) {
      overflow: hidden;
    }

    &:not(.bottom.current) {
      top: 0;
    }

    &.top {
      backface-visibility: hidden;
      background: var(--flipping-number-background-color-top);
      border-top-left-radius: 10px;
      border-top-right-radius: 10px;
      box-shadow: var(--flipping-number-box-shadow-top);
      height: 50%;

      &.next {
        &:before {
          animation: ${({
            animationDuration,
            shouldAnimate,
          }: NumberContainerProps) =>
            shouldAnimate
              ? css`
                  ${show} ${animationDuration /
                  2}ms ease-in reverse forwards ${animationDuration / 2}ms
                `
              : "none"};
          background: linear-gradient(
            to bottom,
            rgba(0, 0, 0, 0.1) 0%,
            rgba(0, 0, 0, 1) 100%
          );
          content: "";
          height: 100%;
          left: 0;
          position: absolute;
          top: 0;
          width: 100%;
        }
      }
    }

    &.bottom {
      align-items: flex-end;
      background: var(--flipping-number-background-color-bottom);
      border-bottom-left-radius: 10px;
      border-bottom-right-radius: 10px;
      box-shadow: var(--flipping-number-box-shadow-bottom);

      &.current {
        height: 50%;
        border-top: 2px solid black;
        bottom: 0;
        transform: rotateX(5deg);

        &:before {
          animation: ${({ animationDuration, shouldAnimate }) =>
            shouldAnimate
              ? css`
                  ${show} ${animationDuration * 0.85}ms ease-in forwards
                `
              : "none"};
          background: linear-gradient(
            to top,
            rgba(0, 0, 0, 0.1) 0%,
            rgba(0, 0, 0, 1) 100%
          );
          content: "";
          height: 100%;
          left: 0;
          opacity: 0;
          position: absolute;
          top: 0;
          width: 100%;
        }
      }
    }

    &.flipper {
      animation: ${({ animationDuration, shouldAnimate }) =>
        shouldAnimate
          ? css`
              ${flip} ${animationDuration}ms ease-in forwards
            `
          : "none"};
      height: 50%;
      transform-origin: bottom;
      transform-style: preserve-3d;

      .bottom,
      .top {
        backface-visibility: hidden;
        height: 100%;
      }

      .bottom {
        transform: rotateX(0.5turn);
      }

      .top {
        transform: rotateX(0deg);

        &:before {
          animation: ${({
            animationDuration,
            shouldAnimate,
          }: NumberContainerProps) =>
            shouldAnimate
              ? css`
                  ${show} ${animationDuration}ms ease-in forwards
                `
              : "none"};
          background: linear-gradient(
            to bottom,
            rgba(0, 0, 0, 0.1) 0%,
            rgba(0, 0, 0, 1) 100%
          );
          content: "";
          height: 100%;
          left: 0;
          opacity: 0;
          position: absolute;
          top: 0;
          width: 100%;
          z-index: 100;
        }
      }
    }
  }
`;
