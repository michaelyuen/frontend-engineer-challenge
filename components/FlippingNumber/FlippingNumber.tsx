import { useEffect, useRef } from "react";
import { FlippingNumberContainer, NumberContainer } from "./styles";

export interface FlippingNumberProps {
  animationDuration?: number;
  label?: string;
  number?: number;
}

export const FlippingNumber: React.FC<FlippingNumberProps> = ({
  animationDuration = 1000,
  label,
  number,
}) => {
  const previousNumberRef = useRef(number);
  const { current: previousNumber } = previousNumberRef;

  useEffect(() => {
    previousNumberRef.current = number;
  }, [number]);

  return (
    <FlippingNumberContainer>
      <NumberContainer
        animationDuration={animationDuration}
        key={number}
        shouldAnimate={previousNumber !== number}
      >
        <div className="top next">{number}</div>
        <div className="flipper">
          <div className="bottom">{number}</div>
          <div className="top">{previousNumber}</div>
        </div>
        <div className="bottom current">{previousNumber}</div>
      </NumberContainer>
      {label && <p>{label}</p>}
    </FlippingNumberContainer>
  );
};
