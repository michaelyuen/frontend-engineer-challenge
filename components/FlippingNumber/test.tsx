import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import { FlippingNumber } from "./FlippingNumber";

describe("FlippingNumber", () => {
  test("renders", () => {
    render(<FlippingNumber animationDuration={0} label="Minutes" number={1} />);
    expect(screen.getByText("Minutes")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <FlippingNumber animationDuration={0} label="Seconds" number={59} />
    );
    expect(container).toMatchSnapshot();
  });

  test("has no a11y violations", async () => {
    const { container } = render(
      <FlippingNumber animationDuration={0} label="Hours" number={0} />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
