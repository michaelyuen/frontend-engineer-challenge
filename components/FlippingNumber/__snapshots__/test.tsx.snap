// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlippingNumber matches snapshot 1`] = `
<div>
  <div
    class="sc-bcXHqe jPLVIK"
  >
    <div
      class="sc-gswNZR iswcfZ"
    >
      <div
        class="top next"
      >
        59
      </div>
      <div
        class="flipper"
      >
        <div
          class="bottom"
        >
          59
        </div>
        <div
          class="top"
        >
          59
        </div>
      </div>
      <div
        class="bottom current"
      >
        59
      </div>
    </div>
    <p>
      Seconds
    </p>
  </div>
</div>
`;
